### PR TITLE
allow re-use of existing pods for restarts

### PR DESCRIPTION
**Pull request by scoheb**
[JENKINS-47476](https://issues.jenkins-ci.org/browse/JENKINS-47476)

[FIXES JENKINS-47476]

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 1]
```diff
@@ -74,10 +74,10 @@
     private static final String JENKINS_HOME = "JENKINS_HOME=";
     private static final Logger LOGGER = Logger.getLogger(ContainerExecDecorator.class.getName());
 
-    private final transient KubernetesClient client;
+    private transient KubernetesClient client;
 
     @SuppressFBWarnings(value = "SE_TRANSIENT_FIELD_NOT_RESTORED", justification = "not needed on deserialization")
-    private final transient List<Closeable> closables = new ArrayList<>();
```
Comment (iocanel) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Why?

Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: When the pipeline is resumed, since it is transient, it will not be deserialized...therefore when we come to add to array, it will NPE...therefore, we check before we add to it and create if necessary. Therefore it cannot be final.

Comment (iocanel) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Makes sense to me.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java [hunk 2]
```diff
@@ -40,7 +46,9 @@
     @Override
     public boolean start() throws Exception {
 
-        Cloud cloud = Jenkins.getInstance().getCloud(step.getCloud());
+        LOGGER.info("starting");
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java]: can you make this more descriptive and probably just `FINE` level ?

Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java]: I will remove this.

File: pom.xml [hunk 3]
```diff
@@ -43,7 +43,7 @@
     <!-- jenkins plugins versions -->
     <jenkins-basic-steps.version>2.3</jenkins-basic-steps.version>
     <jenkins-credentials.version>2.1.7</jenkins-credentials.version>
-    <jenkins-durable-task.version>1.13</jenkins-durable-task.version>
+    <jenkins-durable-task.version>1.16</jenkins-durable-task.version>
```
[the remark this answers is not in the document]
Comment (scoheb) [reply pom.xml]: Yes. I am waiting for the new release that will include https://github.com/jenkinsci/durable-task-plugin/pull/49

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 4]
```diff
@@ -129,10 +129,18 @@ public void launch(SlaveComputer computer, TaskListener listener) {
             String podId = pod.getMetadata().getName();
             String namespace = StringUtils.defaultIfBlank(slave.getNamespace(), client.getNamespace());
 
-            LOGGER.log(Level.FINE, "Creating Pod: {0} in namespace {1}", new Object[]{podId, namespace});
-            pod = client.pods().inNamespace(namespace).create(pod);
-            LOGGER.log(Level.INFO, "Created Pod: {0} in namespace {1}", new Object[]{podId, namespace});
-            listener.getLogger().printf("Created Pod: %s in namespace %s", podId, namespace);
+            Pod checkPod = client.pods().inNamespace(namespace).withName(podId).get();
```
[the remark this answers is not in the document]
Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: A simple test (https://paste.fedoraproject.org/paste/rwMXdUC670i5PIZFAokjlQ) using the current master gives this in the logs when a restart happens:

https://paste.fedoraproject.org/paste/BRTBZEMvyPxsOo2BJJf6uw

With or without idleMinutes

File: src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithLongSleep.groovy [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        containerTemplate(name: 'busybox', image: 'startx/fedora', ttyEnabled: true, command: '/bin/cat'),
+]) {
+
+    node ('mypod') {
+        stage('Run') {
+            container('busybox') {
+                    semaphore 'wait'
+                    sh '''
+                    #!/bin/bash
+                    for i in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
```
Comment (carlossg) [inline src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithLongSleep.groovy]: sh 'for i in `seq 1 20`; do echo $i; sleep 5; done'

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (carlossg) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: interesting, this was changed in abb2b63b7c0ada20fb2fb0b98337755daae6bc72, can you take a look @vlatombe

File: src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithLongSleep.groovy (hunk 5 above)
Comment (scoheb) [reply src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithLongSleep.groovy]: I was having issues with 'seq' at one time...I thought it was due to the busybox image. I will update my sh call.

File: pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jglick) [reply pom.xml]: Use `1.16-20171108.031054-1` in the meantime. Based on https://github.com/jenkinsci/durable-task-plugin/pull/49/commits/9b0b7234b56b659832168b063ac4b18e396c6087.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java [hunk 6]
```diff
@@ -33,9 +34,22 @@
...
 
+    @Override
+    public void onResume() {
+        super.onResume();
+        LOGGER.log(Level.FINE, "ContainerLogStepExecution onResume");
+        try {
+            KubernetesNodeContext nodeContext = new KubernetesNodeContext(getContext());
+            client = nodeContext.connectToCloud();
+        } catch (Exception e) {
+            ContainerLogStepExecution.this.getContext().onFailure(e);
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java]: Include a TODO comment mentioning [JENKINS-40161](https://issues.jenkins-ci.org/browse/JENKINS-40161).

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java [hunk 7]
```diff
@@ -24,10 +24,24 @@
...
+    @Override
+    public void onResume() {
+        super.onResume();
+        LOGGER.log(Level.FINE, "ContainerStepExecution onResume");
+        try {
+            KubernetesNodeContext nodeContext = new KubernetesNodeContext(getContext());
+            client = nodeContext.connectToCloud();
+            decorator.setKubernetesClient(client);
+        } catch (Exception e) {
+            ContainerStepExecution.this.getContext().onFailure(e);
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java]: ditto

File: src/test/java/org/jvnet/hudson/test/RestartableJenkinsNonLocalhostRule.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        HudsonHomeLoader loader = new HudsonHomeLoader() {
+            @Override
+            public File allocate() throws Exception {
+                return home;
+            }
+        };
+
+        // run each step inside its own JenkinsRule
+        for (Statement step : steps) {
+            j = new JenkinsRuleNonLocalhost().with(loader);
```
Comment (jglick) [inline src/test/java/org/jvnet/hudson/test/RestartableJenkinsNonLocalhostRule.java]: You could probably use https://github.com/jenkinsci/jenkins-test-harness/pull/75 and avoid a lot of code duplication, no?

File: src/test/java/org/jvnet/hudson/test/JenkinsRuleNonLocalhost.java [hunk 9]
```diff
@@ -42,8 +40,6 @@
 import org.eclipse.jetty.webapp.Configuration;
 import org.eclipse.jetty.webapp.WebAppContext;
 import org.eclipse.jetty.webapp.WebXmlConfiguration;
-import org.jvnet.hudson.test.JenkinsRule;
-import org.jvnet.hudson.test.ThreadPoolImpl;
```
Comment (jglick) [inline src/test/java/org/jvnet/hudson/test/JenkinsRuleNonLocalhost.java]: Why these changes?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: I guess a simple check of `launched` value at the beginning of the method would be enough. If the pod has been deleted in the mean time we may risk launching another one, leading to an inconsistent state anyway.

[the remark this answers is not in the document]
Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: @Vlatombe so you are ok with the approach in this PR?

[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: I think the check pod logic can be replaced by a simple check for `launched==true`. The launcher is meant to be called only once. On restart we just expect the agent to connect back.

[the remark this answers is not in the document]
Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: @Vlatombe  but how is `launched` being persisted across restarts? on a restart, the attempt to reconnect fails with a handshake error:

https://paste.fedoraproject.org/paste/yHv3FoJ35r7vG8J3hWwYEQ

File: src/test/java/org/jvnet/hudson/test/JenkinsRuleNonLocalhost.java (hunk 9 above)
Comment (scoheb) [reply src/test/java/org/jvnet/hudson/test/JenkinsRuleNonLocalhost.java]: IntelliJ showed them as unused.

File: pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (scoheb) [reply pom.xml]: Thanks @jglick

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java (hunk 6 above)
Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java]: Done

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java (hunk 7 above)
Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java]: Done

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: @scoheb Should be in the slave's config.xml > launcher > launched.

Dunno about the handshake for now since the real error is likely in the master's logs.

[the remark this answers is not in the document]
Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: I checked the slave config.xml once launched and while running a build:

https://paste.fedoraproject.org/paste/cp~cnrkWdoDkoZS~OLkWLg

Looks like the launched property (when true) is not saved. Do we need a `slave.save()` after at the end of the `launch()` ?

[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: Sounds like the right move.

[the remark this answers is not in the document]
Comment (scoheb) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: @Vlatombe  

With master and ensuring 'launched == true' in config.xml (via direct save() in launch()), here is Jenkins log after restart:
https://paste.fedoraproject.org/paste/UhqADZdp010vXMAk2HI8nw

No trace of why the slave is not connecting properly...

**Comment by scoheb** [comment]
@Vlatombe Since we were not checking `launched`, the provision would fail (since pod already existed) and it would fail into `catch` block where we would `removeNode()`, therefore causing the subsequent  handshake error on the agent side. This become more evident when I switched to using JNLP-4 and it clearly said it was rejecting the client since the node name was not found.
Thanks for your help!

**Comment by scoheb** [comment]
@carlossg Would you be able to provide some of the test logs from https://jenkins.gke.csanchez.org/job/jenkinsci/job/kubernetes-plugin/job/PR-244/10/ ?

I can run the `RestartPipelineTest` locally using minikube and `-Djenkins.host.address=192.168.99.1` but not in the test environment. Surprisingly enough,  `KubernetesPipelineTest` runs fine above, so I must not be setting up RestartPipelineTest` correctly.

File: src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithLongSleep.groovy [hunk 10]
```diff
@@ -0,0 +1,17 @@
+package org.csanchez.jenkins.plugins.kubernetes.pipeline
+
+podTemplate(label: 'mypod', containers: [
+        containerTemplate(name: 'busybox', image: 'startx/fedora', ttyEnabled: true, command: '/bin/cat'),
```
Comment (carlossg) [inline src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithLongSleep.groovy]: why not `busybox` image ?

File: src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithMultipleContainerCalls.groovy [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package org.csanchez.jenkins.plugins.kubernetes.pipeline
+
+podTemplate(label: 'mypod', namespace: 'default', idleMinutes: 0, containers: [
+        containerTemplate(name: 'busybox', image: 'startx/fedora', ttyEnabled: true, command: '/bin/cat'),
```
Comment (carlossg) [inline src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runInPodWithRestartWithMultipleContainerCalls.groovy]: why not busybox image ?

**Comment by scoheb** [comment]
@carlossg  updated podTemplates in *.groovy to use `busybox`. At one time, I thought the `ps` output from the busybox images was causing issues...

If the tests fail again, can you help me and provide any logs from k8s side?
Or maybe update the Jenkinsfile that is used to ALWAYS capture the junit results and logs...it is not the case now.

**Comment by carlossg** [comment]
now test logs are captured and removed the explicit namespace definition, that was the reason for failure, let's see

**Comment by scoheb** [comment]
@carlossg  Thanks

Seems like tests are failing due to leftover pods...can you take a look please?

**Comment by scoheb** [comment]
@carlossg Thanks for your test help!

**Comment by carlossg** [comment]
This looks good, tests are passing. 
Would love a review from someone with enough pipeline knowledge like @jglick though

It can't be merged until durable task 1.16 is released

**Comment by scoheb** [comment]
@jglick  has released 1.16 already...I have amended this PR to include it.

**Comment by carlossg** [comment]
ok, last ask hopefully :) can you add license headers to the new java files you added, similar to the other ones?

**Comment by scoheb** [comment]
@carlossg it would be a pleasure!

**Comment by scoheb** [comment]
@carlossg added to `java` files only. Do we need to add to `.groovy` files too?

**Comment by jglick** [comment]
BTW as a matter of courtesy to reviewers, it is nice to hyperlink to JIRA [JENKINS-47476](https://issues.jenkins-ci.org/browse/JENKINS-47476) in the PR description, and then also mark the JIRA **In Review**.

**Comment by jglick** [comment]
(assuming this PR purports to fix the defect)

**Comment by scoheb** [comment]
@jglick Thanks. I will properly link to the JIRA and mark it as In Review in the future.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java [hunk 12]
```diff
@@ -24,10 +25,26 @@
 
     private static final transient Logger LOGGER = Logger.getLogger(ContainerStepExecution.class.getName());
 
-    private final ContainerStep step;
+    @Inject(optional=true)
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java]: Again, no, you are not `Inject`ing anything. This is a little more confusing since you are extending `AbstractStepExecutionImpl`, which _can_ support Guice, but (as of #124) using the nondeprecated `super` constructor overload which does not.

Whether there is actually any reason to be extending `AbstractStepExecutionImpl` is another question. Normally you would just extend `StepExecution` directly. The only reason to continue to extend `AbstractStepExecutionImpl` is to maintain compatibility of `program.dat` deserialization across restarts when the plugin is updated—Java Serialization treats a change of superclass as unconditionally incompatible, even when `serialVersionUID` is maintained. But if I understand correctly the issue that this whole PR is addressing, it is that this step never worked across restarts to begin with! Thus there can be no _new_ problem introduced by getting rid of `AbstractStepExecutionImpl`.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java [hunk 13]
```diff
@@ -30,7 +33,12 @@
 
     private static final transient String NAME_FORMAT = "%s-%s";
 
-    private final PodTemplateStep step;
+    @Inject(optional=true)
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java]: ditto

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java [hunk 14]
```diff
@@ -30,7 +33,12 @@
...
+    @SuppressFBWarnings(value = "SE_TRANSIENT_FIELD_NOT_RESTORED", justification = "not needed on deserialization")
+    private final transient PodTemplateStep step;
+    private String cloudName;
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java]: If this is all we needed from the `step`, perhaps this did not need to be a field at all. Just initialize `cloudName` in the constructor, and make it `final`.

BTW if you are introducing support for resuming these steps for the first time, you can just make the `serialVersionUID` be, say, `1`. Hereafter you need to be careful to maintain the SVUID, and also to consider the effect of deserializing and old instance—for example, new fields might be null and you need to handle that possibility.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+    @Rule
+    public RestartableJenkinsNonLocalhostRule story = new RestartableJenkinsNonLocalhostRule(44000);
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public LoggerRule logs = new LoggerRule().record(Logger.getLogger(KubernetesCloud.class.getPackage().getName()),
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: Feel free to file a PR for an overload `record(Package)` (or `recordPackage(Class<?>)`), which would be useful on occasion.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+                        Collections.<NodeProperty<?>>emptyList())); // TODO JENKINS-26398 clumsy
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(loadPipelineScript("runInPodWithRestartWithMultipleContainerCalls.groovy")
+                        , true));
+                WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b);
+                SemaphoreStep.success("wait/1", false);
+                // we need to wait until we are sure that the sh
+                // step has started...
+                Thread.sleep(10000);
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: Flaky. Perhaps:

```java
r.waitForMessage("+ sleep 5", b);
```

?

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+            }
+        });
+        story.addStep(new Statement() {
+            @SuppressWarnings("SleepWhileInLoop")
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: ?

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+
+                story.j.jenkins.addNode(new DumbSlave("slave", "dummy", tmp.newFolder("remoteFS").getPath(), "1",
+                        Node.Mode.NORMAL, "", new JNLPLauncher(), RetentionStrategy.NOOP,
+                        Collections.<NodeProperty<?>>emptyList())); // TODO JENKINS-26398 clumsy
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(loadPipelineScript("runInPodWithRestartWithMultipleContainerCalls.groovy")
+                        , true));
+                WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b);
+                SemaphoreStep.success("wait/1", false);
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: Makes little sense to wait for a `semaphore` and then immediately let it proceed anyway. What you are actually waiting for is something else, so just drop the `semaphore`.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+
+                story.j.jenkins.addNode(new DumbSlave("slave", "dummy", tmp.newFolder("remoteFS").getPath(), "1",
+                        Node.Mode.NORMAL, "", new JNLPLauncher(), RetentionStrategy.NOOP,
+                        Collections.<NodeProperty<?>>emptyList())); // TODO JENKINS-26398 clumsy
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(loadPipelineScript("runInPodWithRestartWithLongSleep.groovy")
+                        , true));
+                WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b);
+                SemaphoreStep.success("wait/1", false);
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: ditto

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+                // we need to wait until we are sure that the sh
+                // step has started...
+                Thread.sleep(10000);
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: ditto

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+
+                story.j.jenkins.addNode(new DumbSlave("slave", "dummy", tmp.newFolder("remoteFS").getPath(), "1",
+                        Node.Mode.NORMAL, "", new JNLPLauncher(), RetentionStrategy.NOOP,
+                        Collections.<NodeProperty<?>>emptyList())); // TODO JENKINS-26398 clumsy
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(loadPipelineScript("getContainerLogWithRestart.groovy")
+                        , true));
+                WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b);
+                SemaphoreStep.success("wait/1", false);
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: ditto

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+                Thread.sleep(1000);
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: ditto

File: src/test/java/org/jvnet/hudson/test/JenkinsRuleNonLocalhost.java [hunk 23]
```diff
@@ -24,8 +24,6 @@
 
 package org.jvnet.hudson.test;
```
Comment (jglick) [inline src/test/java/org/jvnet/hudson/test/JenkinsRuleNonLocalhost.java]: Eek! Cf. https://github.com/jenkinsci/jenkins-test-harness/pull/84

File: src/test/java/org/jvnet/hudson/test/RestartableJenkinsNonLocalhostRule.java [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+ */
+public class RestartableJenkinsNonLocalhostRule extends RestartableJenkinsRule {
+    private final int port;
+
+    public RestartableJenkinsNonLocalhostRule(int port) {
+        this.port = port;
+    }
+
+    @Override
+    protected JenkinsRule createJenkinsRule(Description description) {
```
Comment (jglick) [inline src/test/java/org/jvnet/hudson/test/RestartableJenkinsNonLocalhostRule.java]: Ah good to see https://github.com/jenkinsci/jenkins-test-harness/pull/75 getting used at least.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 25]
```diff
@@ -120,6 +120,12 @@ public void launch(SlaveComputer computer, TaskListener listener) {
         if (slave == null) {
             throw new IllegalStateException("Node has been removed, cannot launch " + computer.getName());
         }
+        if (launched) {
+            LOGGER.info("Agent: " + slave.getNodeName() + " has already been launched. Activating...");
```
[the remark this answers is not in the document]
Comment (carlossg) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java]: yep, I'd also prefer

    LOGGER.log(INFO, "Agent has already been launched, activating: {}", slave.getNodeName());

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java (hunk 15 above)
Comment (carlossg) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/RestartPipelineTest.java]: I will

**Comment by scoheb** [comment]
@carlossg I believe I have addressed all of Jesse's comments.

**Comment by scoheb** [comment]
Latest test run shows this in logs:

`Failed to count the # of live instances on Kubernetes
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/default/pods?labelSelector=jenkins%3Dslave. Message: Forbidden!Configured service account does`

and

`org.jvnet.hudson.reactor.ReactorException: java.lang.Error: java.lang.reflect.InvocationTargetException
	at org.jvnet.hudson.reactor.Reactor.execute(Reactor.java:269)
	at jenkins.InitReactorRunner.run(InitReactorRunner.java:47)
	at jenkins.model.Jenkins.executeReactor(Jenkins.java:1110)
	at jenkins.model.Jenkins.<init>(Jenkins.java:926)
	at hudson.model.Hudson.<init>(Hudson.java:85)
	at org.jvnet.hudson.test.JenkinsRule.newHudson(JenkinsRule.java:611)
	at org.jvnet.hudson.test.JenkinsRule.before(JenkinsRule.java:384)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:537)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.Error: java.lang.reflect.InvocationTargetException
	at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:110)
	at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:175)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:282)
	at jenkins.model.Jenkins$7.runTask(Jenkins.java:1086)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:210)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	... 1 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor323.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:104)
	... 8 more
Caused by: java.lang.IndexOutOfBoundsException: Index: 0
	at java.util.Collections$EmptyList.get(Collections.java:4454)
	at hudson.ExtensionList.get(ExtensionList.java:172)
	at org.jenkinsci.plugins.pipeline.modeldefinition.DescriptorLookupCache.getPublicCache(DescriptorLookupCache.java:53)
	at org.jenkinsci.plugins.pipeline.modeldefinition.DescriptorLookupCache.invalidateGlobalCache(DescriptorLookupCache.java:58)`

**Comment by scoheb** [comment]
@carlossg Still getting this warning which I think is causing tests to fail...:

WARNING	o.c.j.p.k.KubernetesCloud#provision: Failed to count the # of live instances on Kubernetes
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/default/pods?labelSelector=jenkins%3Dslave. Message: Forbidden!Configured service account does....

**Comment by scoheb** [comment]
@carlossg  Sorry. my bad. Some of the test groovy files had a namespace defined in them...I removed them and now tests are passing.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java [hunk 26]
```diff
@@ -33,9 +35,22 @@
...
     private static final transient Logger LOGGER = Logger.getLogger(ContainerLogStepExecution.class.getName());
 
-    private final ContainerLogStep step;
+    @SuppressFBWarnings(value = "SE_TRANSIENT_FIELD_NOT_RESTORED", justification = "not needed on deserialization")
+    private transient final ContainerLogStep step;
     private transient KubernetesClient client;
 
+    @Override
+    // TODO Revisit for JENKINS-40161
+    public void onResume() {
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java]: I'm not convinced this is the proper implementation for this step. The parent class `SynchronousNonBlockingStepExecution` makes it clear that resuming such a step is unsupported, the bulk of the execution being run in a different thread, which is not resumed on restart.

As I understand it, the current `onResume` is just a no-op, so you may as well remove injection of 'client' as it won't be used in later calls.

Comment (carlossg) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java]: @jglick you didn't raise this point before, do you agree with @Vlatombe  ?

**Comment by scoheb** [comment]
I agree with you @Vlatombe ... I was just going for completeness (i.e. to have all pipeline steps resume) but as I look at it, the step is not a long running one like container()

I will amend my PR

**Comment by scoheb** [comment]
@Vlatombe  @carlossg  reverted the change to `ContainerLogStepExecution`. Please take a look.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java [hunk 27]
```diff
@@ -16,6 +16,8 @@
 
 package org.csanchez.jenkins.plugins.kubernetes.pipeline;
 
+import com.google.inject.Inject;
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerLogStepExecution.java]: Some unused imports to clean up here.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java [hunk 28]
```diff
@@ -4,30 +4,47 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import com.google.inject.Inject;
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java]: Ditto.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java [hunk 29]
```diff
@@ -5,12 +5,15 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import com.google.inject.Inject;
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/PodTemplateStepExecution.java]: Ditto

**Comment by carlossg** [comment]
tests passing so 🎉

thanks for the hard work!

**Comment by chancez** [comment]
This doesn't seem to have fixed anything for me. Jenkins is still rejecting old agents because it's an unknown client name.

**Comment by chancez** [comment]
It turns out my plugin upgrade got reverted while I was testing by someone. This is working great!
